model/riscv_insts_zks.sail

```diff
@@ -0,0 +1,92 @@
+/*
+ * This file contains instructions whose definitions are independent of XLEN.
+ */
+
+
+/*
+ * Scalar Cryptography Extension - Scalar SM3 instructions
+ * ----------------------------------------------------------------------
+ */
+
+/* AST Types for SM3.p0/1 */
+union clause ast = SM3P0 : (regidx, regidx)
+union clause ast = SM3P1 : (regidx, regidx)
+
+/* Decode clause for sm3.p0 */
+mapping clause encdec = SM3P0 (rs1, rd) 
+  <-> 0b00 @ 0b01000 @ 0b01000 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+/* Decode clause for sm3.p1 */
+mapping clause encdec = SM3P1 (rs1, rd) 
+  <-> 0b00 @ 0b01000 @ 0b01001 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+/* Asssembly code mapping for sm3p0 */
+mapping clause assembly = SM3P0 (rs1, rd) <->
+  "sm3p0" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+/* Asssembly code mapping for sm3p1 */
+mapping clause assembly = SM3P1 (rs1, rd) <->
+  "sm3p1" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+function clause execute (SM3P0(rs1, rd)) = {
+  let r1     : bits(32) = X(rs1)[31..0];
+  let result : bits(32) =  r1 ^ rol32(r1,  9) ^ rol32(r1, 17);
+  X(rd) = EXTS(result);
+  RETIRE_SUCCESS
+}
+
+function clause execute (SM3P1(rs1, rd)) = {
+  let r1     : bits(32) = X(rs1)[31..0];
+  let result : bits(32) =  r1 ^ rol32(r1, 15) ^ rol32(r1, 23);
+  X(rd) = EXTS(result);
+  RETIRE_SUCCESS
+}
+
+/*
+ * Scalar Cryptography Extension - Scalar SM4 instructions
+ * ----------------------------------------------------------------------
+ */
+
+/* AST clauses for the SM4 instructions. */
+union clause ast = SM4ED : (bits(2),regidx,regidx,regidx)
+union clause ast = SM4KS : (bits(2),regidx,regidx,regidx)
+
+/* Encode / decode mappings for the SM4 instructions. */
+mapping clause encdec = SM4ED   (bs,rs2,rs1,rd) 
+    <-> bs @ 0b11000 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0110011
+
+mapping clause encdec = SM4KS   (bs,rs2,rs1,rd) 
+    <-> bs @ 0b11010 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0110011
+
+/* Assembly string mappings for the SM4 instructions. */
+mapping clause assembly = SM4ED (bs,rs2,rs1,rd) <->
+    "sm4ed" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2) ^ sep() ^ hex_bits_2(bs)
+
+mapping clause assembly = SM4KS (bs,rs2,rs1,rd) <->
+    "sm4ks" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2) ^ sep() ^ hex_bits_2(bs)
+
+/* Execute clause for the sm4 instructions. */
+function clause execute (SM4ED (bs,rs2,rs1,rd)) = {
```

Comment:
Yeah, it's because the instructions use what's called a "T-Tables" based implementation strategy for SM4 (as with the 32-bit AES instructions) so there is no expectation that the Sail code would correspond to the original algorithm specification. I can add a note the description of the instructions for explaining this certainly. Unfortunately, I've found this is quite common in cryptography, where the "naive" algorithm implementation / specification looks absolutely nothing like the actual real world implementation.